dev/doc/SProp.md

```diff
@@ -0,0 +1,71 @@
+Irrelevant`.
+
+This annotation is done by transforming the binder name `'a` into a
+`'a Context.binder_annot = { binder_name : 'a; binder_relevance :
+Sorts.relevance }`, eg `Prod of Name.t * types * types` becomes `Prod
+of Name.t Context.binder_annot * types * types`.
+
+If you just carry binder names around without looking at them no
+change is needed, eg if you have `match foo with Lambda (x, a, b) ->
+Prod (x, a, type_of (push_rel (LocalAssum (x,a)) env) b)`. Otherwise
+see `context.mli` for a few combinators on the `binder_annot` type.
+
+As a plugin you can generally just use `Relevant` everywhere, the
+kernel will fix it if needed when it checks the terms you produce. The
+only issue is that if you generate `Relevant` when it should have been
+`Irrelevant` you won't be able to use proof irrelevance on that
+variable until the kernel fixes it. See refman for examples as Coq
+also uses `Relevant` incorrectly in some places.
+
+When making `Relevant` annotations you can use some convenience
+functions from `Context` (eg `annotR x = make_annot x Relevant`), also
+`mkArrowR` from `Constr`/`EConstr` which has the signature of the old
+`mkArrow`.
+
+You can enable the debug warning `bad-relevance` to help find places
+where you generate incorrect annotations.
+
+Relevance can be inferred from a well-typed term using functions in
+`Retypeops` (for `Constr`) and `Retyping` (for `EConstr`). For `x` a
+term, note the difference between its relevance as a term (is `x :
+(_ : SProp)`) and as a type (is `x : SProp`), there are functions for
+both kinds.
+
+## Case annotations
+
+`Constr.case_info` now contains a relevance annotation, this is the
+relevance of the whole case term (eg `match ... return _ : SProp
+with ... end` has `ci_relevance = Irrelevant`).
+
+Additionally, the `Case` constructor has an additional `constr option`
+argument. When eliminating an irrelevant inductive to a relevant value
+it contains the type of the discriminee. It is usually bound to `is`
+(for `indices` which it is meant to provide).
+
+ie
+- `match ... : (foo : SProp) return ... : SProp` => `is = None`
+- `match ... : (foo : Prop/Set/Type) return ... : SProp/Prop/Set/Type` => `is = None`
+- `match ... : (foo : SProp) return ... : Prop/Set/Type` => `is = Some foo`
+
+Using `Relevant`/`None` everywhere is backward compatible but the
+kernel is not (yet?) able to repair incorrect case annotations so your
+code may fail for `SProp` users.
+
+In general you should probably be using
+`Inductiveops.make_case_or_project` instead of producing the `is`
+annotation yourself, this will also handle primitive records for you.
+You still need to pass it the right `case_info` and so the right
+`ci_relevance` though.
+
+NB: make_case_or_project is already passed an env, maybe we should
+have it retype to get `ci_relevance`. However currently the env is
+only for globals so this will break those who call it with incorrect
+rel_context (IIRC cases calls with incorrect rel_context).
```

Comment:
I assume this "IIRC" is a todo.

Reply:
The entire paragraph is a todo.

Reply:
ie something to consider after the PR is merged